modules/library/metadata/src/main/java/org/geotools/util/Utilities.java

```diff
@@ -755,4 +760,16 @@ public static void ensureArgumentNonNull(final String name, final Object object)
                 Function.identity(),
                 (first, second) -> second);
     }
+
+    /**
+     * Asserting if the passed string is not prefixed with '..' to allow an attacker to traverse
+     * outside the referenced directory
+     *
+     * @param fileName to checked for zip sip vulnerability
+     * @throws IOException if passed fileName is found to be vulnerable to zlip slip, e.g fileName
+     *     should not be prefixed with '..'
+     */
+    public static void assertNotZipSlipVulnerable(String fileName) throws IOException {
+        if (fileName.startsWith(ZIP_SLIP_EXPRESSION)) throw new IOException("Vulnerable File Name");
```

Comment:
Is this really a sufficient check? What happens if I pass `./../../something.important` ?

Reply:
so it needs to look for any ".." inside the file name, right?

Reply:
I'd say one needs a "..<fileSeparator>" to trigger it, anywhere in the file name... or not?

Reply:
this check is insufficient; the filename should be canonicalized and then the check for starting with ..%s should be done.
It requires creating a File object and then checking the path.

```java
 File destinationfile = new File(fileName);
 String canonicalDestinationFile = destinationfile.getCanonicalPath();
 if (canonicalDestinationFile.startsWith(ZIP_SLIP_EXPRESSION)) {
   throw new IOException("Vulnerable File Name");
 }
```

In reality the fileName should be compared to the directory where the zipfile should be expanded and not just for starting with `..%s`, see the code at: https://snyk.io/research/zip-slip-vulnerability

Reply:
I think @aaime has proposed a simple check, ".." means reverse traversal.

Reply:
The approach suggested by @mprins looks good to me (check the page he linked and the code he pasted)

Reply:
> this check is insufficient; the filename should be canonicalized and then the check for starting with ..%s should be done.
> It requires creating a File object and then checking the path.
>  File destinationfile = new File(fileName);
>  String canonicalDestinationFile = destinationfile.getCanonicalPath();
>  if (canonicalDestinationFile.startsWith(ZIP_SLIP_EXPRESSION)) {
>    throw new IOException("Vulnerable File Name");
>  }
> In reality the fileName should be compared to the directory where the zipfile should be expanded and not just for starting with ..%s, see the code at: https://snyk.io/research/zip-slip-vulnerability

this looks simple and great, I will update accordingly. Thanks

Reply:
@mprins can you check again?